eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Tick.hs

```diff
@@ -163,6 +168,64 @@ validatingTickTransition nes slot = do
 
   pure $ nes' {nesEs = es''}
 
+-- | This is a limited version of 'validatingTickTransition' which is only suitable
+-- for the future ledger view.
+validatingTickTransitionFORECAST ::
+  forall tick era.
+  ( State (tick era) ~ NewEpochState era,
+    State (Core.EraRule "PPUP" era) ~ PPUPState era,
+    UsesPParams era,
+
+    BaseM (tick era) ~ ShelleyBase,
+    STS (tick era)
+  ) =>
+  NewEpochState era ->
+  SlotNo ->
+  TransitionRule (tick era)
+validatingTickTransitionFORECAST nes slot = do
+  -- This whole function is a specialization of an inlined 'NEWEPOCH'.
+  --
+  -- The ledger view is built entirely from the 'nesPd' and 'esPP' and
+  -- '_genDelegs', so those are the only pieces we bother to update here.
+
+  epoch <- liftSTS $ do
+    ei <- asks epochInfoPure
+    epochInfoEpoch ei slot
+
+  let es = nesEs nes
+      EpochState
+        { esSnapshots = ss,
+          esLState = ls,
+          esPp = pp
+        } = es
+
+  -- the relevant 'NEWEPOCH' logic
+  let pd' = _pstakeMarkPoolDistr ss
+  if unEpochNo epoch /= unEpochNo (nesEL nes) + 1 then pure nes else do
+
+    -- We can skip 'SNAP'; we already have the equivalent pd'.
+
+    -- We can skip 'MIR', 'POOLREAP', 'UPEC' (which calls 'NEWPP'); we don't
+    -- need to do the checks: if the checks would fail, then the node will fail
+    -- in the 'TICK' rule if it ever tries to validate blocks for which the
+    -- return value here was used to validate their headers.
+
+    coreNodeQuorum <- liftSTS $ asks quorum
+
+    let pp' =
+            maybe pp id
+          $ votedValue
+              (proposals $ _ppups $ lsUTxOState ls)
+              pp
+              (fromIntegral coreNodeQuorum)
+
+    pure $! nes {
+        nesPd = pd'
+      , nesEs = adoptGenesisDelegs
```

Comment:
just for posterity, let me mention that the genesis delegation mapping also needs to be updating in the `e =/= e + 1` case. The re-delegation of the genesis keys is staged for adoption one security window from the time the delegation certificate is submitted (ie not necessarily on an epoch boundary).

hooray for property tests! :tada: